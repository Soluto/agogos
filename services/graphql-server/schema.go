package main

import (
	graphql "github.com/graphql-go/graphql"
	"github.com/vektah/gqlparser"
	"github.com/vektah/gqlparser/ast"

	"context"
	"fmt"
	"google.golang.org/grpc"
	"graphql-gateway/generated"
	"graphql-gateway/utils"
	"io"
	"os"
	"time"
)

func getenv(key, fallback string) string {
	value := os.Getenv(key)
	if len(value) == 0 {
		return fallback
	}
	return value
}

var (
	address = getenv("SCHEMA_REGISTRY_URL", "graphql-gateway.registry:81")
)

type schemaResult struct {
	schema *graphql.Schema
	err    error
}

func subscribeToSchema(schemas chan schemaResult) (err error) {
	defer utils.Recovery(&err)

	conn, err := grpc.Dial(address, grpc.WithInsecure())
	if err != nil {
		fmt.Println("error initiating GRPC channel")
		return err
	}
	defer conn.Close()

	gqlSchemaClient := gqlschema.NewGqlSchemaClient(conn)

	stream, err := subscribe(gqlSchemaClient)
	if err != nil {
		fmt.Println("error subscribing to registry", err)
		return err
	}

	for {
		gqlSchemaMessage, err := stream.Recv()

		if err == io.EOF {
			fmt.Println("got EOF")
			break
		}
		if err != nil {
			fmt.Println("error receiving message")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		astSchema, err := parseSdl(source{
			name: "schema registry sdl",
			sdl:  gqlSchemaMessage.Schema,
		})
		if err != nil {
			fmt.Println("error parsing SDL")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schema, err := ConvertSchema(astSchema)
		if err != nil {
			fmt.Println("error converting schema")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schemas <- schemaResult{
			schema: schema,
			err:    nil,
		}
	}
	return
}

func subscribe(client gqlschema.GqlSchemaClient) (stream gqlschema.GqlSchema_SubscribeClient, err error) {
	for i := 0; i < 3; i++ {
		stream, err = client.Subscribe(context.Background(), &gqlschema.GqlSchemaSubscribeParams{})
		if err == nil {
			break
		}
		if i+1 < 3 {
			time.Sleep(3000 * time.Millisecond)
		}
	}
	return
}

type source struct {
	name string
	sdl  string
}

func parseSdl(s source) (*ast.Schema, error) {
	schema, err := gqlparser.LoadSchema(&ast.Source{
		Name:  s.name,
		Input: s.sdl,
	})

	if err != nil {
		return nil, err
	}

	return schema, nil
}
