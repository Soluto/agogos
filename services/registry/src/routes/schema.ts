import * as express from "express";
import * as bodyParser from "body-parser";
import sources, { defaultSource } from "../sources-config";
import syncSchema$ from "../sync/sync-service";
import { take } from "rxjs/operators";

const app = express();

const getFromSource = async (source: string, kind: string, res: express.Response) => {
    try {
        const gqlObjects = await sources[source].getGqlObjects(kind);
        res.send(gqlObjects);
    } catch (error) {
        console.warn(`Failed to get ${kind} from source - ${source}`, {
            error
        });
        res.sendStatus(500);
    }
};

const postSource = async (
    source: string,
    kind: string,
    name: string,
    definition: string,
    res: express.Response
) => {
    try {
        await sources[source].registerGqlObject(name, kind, definition);
        res.sendStatus(200);
    } catch (error) {
        console.warn(`Failed to register schema to source - ${source}`, {
            name,
            kind,
            error
        });
        res.sendStatus(500);
    }
};

app
    .use(bodyParser.text())
    .get("/:sourceName/:kind", (req, res) => getFromSource(req.params.sourceName, req.params.kind, res))
    .post("/:sourceName/:kind/:name", (req, res) => {
        const { sourceName, kind, name } = req.params;
        const definition = req.body;
        return postSource(sourceName, kind, name, definition, res);
    })
    .get("/", async (_, res) => res.send(await syncSchema$.pipe(take(1)).toPromise()))
    .post(":kind/:name", (req, res) => {
        const { kind, name } = req.params;
        const definition = req.body;
        return postSource(defaultSource, kind, name, definition, res);
    });

export default app;
