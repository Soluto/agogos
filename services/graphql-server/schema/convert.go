package schema

import (
	"agogos/directives/common"
	"agogos/directives/common/exports"
	"agogos/directives/middlewares"
	"agogos/scalars"
	"agogos/server"
	"agogos/utils"
	"fmt"
	"strings"

	graphql "github.com/graphql-go/graphql"
	"github.com/vektah/gqlparser/ast"

	log "github.com/sirupsen/logrus"
)

type schemaContext struct {
	server.ServerContext
	schema     *ast.Schema
	objects    map[string]*graphql.Object
	interfaces map[string]*graphql.Interface
	unions     map[string]*graphql.Union
	enums      map[string]*graphql.Enum
	inputs     map[string]*graphql.InputObject
}

var errMissingResolver = fmt.Errorf("missing resolver")

func convertType(t *ast.Type, c schemaContext) graphql.Type {
	if t.Elem != nil {
		if t.NonNull {
			return graphql.NewNonNull(graphql.NewList(convertType(t.Elem, c)))
		}
		return graphql.NewList(convertType(t.Elem, c))
	} else if t.NonNull {
		return graphql.NewNonNull(convertNamedType(t, c))
	}

	return convertNamedType(t, c)
}

func convertNamedType(t *ast.Type, c schemaContext) graphql.Type {
	switch name := t.Name(); name {
	case "ID":
		return graphql.ID
	case "String":
		return graphql.String
	case "Int":
		return graphql.Int
	case "Float":
		return graphql.Float
	case "Boolean":
		return graphql.Boolean
	case "Date":
		return graphql.DateTime
	case "JSON":
		return scalars.JSON
	default:
		definition := c.schema.Types[name]

		if definition.Kind == ast.Object {
			return convertSchemaObject(definition, c)
		}
		if definition.Kind == ast.InputObject {
			return convertSchemaInputObject(definition, c)
		}
		if definition.Kind == ast.Interface {
			return convertSchemaInterface(definition, c)
		}
		if definition.Kind == ast.Union {
			return convertSchemaUnion(definition, c)
		}
		if definition.Kind == ast.Scalar {
			log.Panic("Custom scalar handling not implemented")
		}
		if definition.Kind == ast.Enum {
			return convertSchemaEnum(definition, c)
		}

		log.Panic("Unreachable code")
		return nil
	}
}

func convertFieldArgs(a ast.ArgumentDefinitionList, c schemaContext) (graphql.FieldConfigArgument, error) {
	args := make(map[string]*graphql.ArgumentConfig)

	for _, d := range a {
		defaultValue, err := d.DefaultValue.Value(make(map[string]interface{}))

		if err != nil {
			return nil, err
		}

		args[d.Name] = &graphql.ArgumentConfig{
			DefaultValue: defaultValue,
			Description:  d.Description,
			Type:         convertType(d.Type, c),
		}
	}

	return args, nil
}

func createResolver(f *ast.FieldDefinition, c schemaContext) middlewares.Resolver {
	resolver := fieldIdentityResolver

	for _, d := range f.Directives {
		middlewareDefinition, ok := common.MiddlewareDefinitions[d.Name]

		if !ok {
			continue
		}

		middleware := middlewareDefinition.CreateMiddleware(c, f, d)

		resolver = middleware.Wrap(resolver)
	}

	return exports.WrapperMiddleware.Wrap(resolver)
}

func fieldIdentityResolver(rp graphql.ResolveParams) (interface{}, error) {
	fieldName := rp.Info.FieldName
	return utils.IdentityResolver(fieldName, rp.Source)
}

func convertSchemaField(f *ast.FieldDefinition, c schemaContext) (*graphql.Field, error) {
	args, err := convertFieldArgs(f.Arguments, c)

	if err != nil {
		return nil, err
	}

	return &graphql.Field{
		Description: f.Description,
		Type:        convertType(f.Type, c),
		Resolve:     createResolver(f, c),
		Args:        args,
	}, nil
}

func convertSchemaEnum(d *ast.Definition, c schemaContext) *graphql.Enum {
	enum, ok := c.enums[d.Name]

	if ok {
		return enum
	}

	values := make(map[string]*graphql.EnumValueConfig)

	for _, v := range d.EnumValues {
		enumValue := new(graphql.EnumValueConfig)
		enumValue.Value = v.Name
		enumValue.Description = v.Description
		values[v.Name] = enumValue
	}

	enum = graphql.NewEnum(graphql.EnumConfig{Name: d.Name, Values: values, Description: d.Description})

	c.enums[d.Name] = enum

	return enum
}

func convertSchemaInputObject(d *ast.Definition, c schemaContext) *graphql.InputObject {
	object, ok := c.inputs[d.Name]

	if ok {
		return object
	}

	fieldsThunk := graphql.InputObjectConfigFieldMapThunk(func() graphql.InputObjectConfigFieldMap {
		fields := make(map[string]*graphql.InputObjectFieldConfig)
		for _, field := range d.Fields {
			if field.Name[:2] == "__" {
				continue
			}

			fields[field.Name] = convertSchemaInputField(field, c)
		}
		return fields
	})

	object = graphql.NewInputObject(graphql.InputObjectConfig{Name: d.Name, Fields: fieldsThunk, Description: d.Description})

	c.inputs[d.Name] = object

	return object
}

func convertSchemaInputField(f *ast.FieldDefinition, c schemaContext) *graphql.InputObjectFieldConfig {
	return &graphql.InputObjectFieldConfig{
		Description: f.Description,
		Type:        convertType(f.Type, c),
	}
}

func convertSchemaUnion(d *ast.Definition, c schemaContext) *graphql.Union {
	union, ok := c.unions[d.Name]

	if ok {
		return union
	}

	types := []*graphql.Object{}

	for _, t := range d.Types {
		types = append(types, convertSchemaObject(c.schema.Types[t], c))
	}

	union = graphql.NewUnion(graphql.UnionConfig{Name: d.Name, Types: types, Description: d.Description})

	c.unions[d.Name] = union

	return union
}

func convertSchemaInterface(d *ast.Definition, c schemaContext) *graphql.Interface {
	object, ok := c.interfaces[d.Name]

	if ok {
		return object
	}

	fieldsThunk := graphql.FieldsThunk(func() graphql.Fields {
		fields := make(map[string]*graphql.Field)
		for _, field := range d.Fields {
			if strings.HasPrefix(field.Name, "__") {
				continue
			}

			schemaField, err := convertSchemaField(field, c)
			if err != nil {
				log.Panic(err)
			}

			fields[field.Name] = schemaField
		}
		return fields
	})

	object = graphql.NewInterface(graphql.InterfaceConfig{Name: d.Name, Fields: fieldsThunk, Description: d.Description})

	c.interfaces[d.Name] = object

	return object
}

func convertSchemaObject(d *ast.Definition, c schemaContext) *graphql.Object {
	if d == nil {
		return nil
	}

	object, ok := c.objects[d.Name]

	if ok {
		return object
	}

	fieldsThunk := graphql.FieldsThunk(func() graphql.Fields {
		fields := make(map[string]*graphql.Field)
		for _, field := range d.Fields {
			if field.Name[:2] == "__" {
				continue
			}

			schemaField, err := convertSchemaField(field, c)
			if err != nil {
				log.Panic(err)
			}

			fields[field.Name] = schemaField
		}
		return fields
	})

	object = graphql.NewObject(graphql.ObjectConfig{Name: d.Name, Fields: fieldsThunk, Description: d.Description})

	c.objects[d.Name] = object

	return object
}

// ConvertSchema converts schema definition to a graphql schema
func convertSchema(serverContext server.ServerContext, astSchema *ast.Schema) (schemaPtr *graphql.Schema, err error) {
	defer utils.Recovery(&err)

	schemaCtx := schemaContext{
		serverContext,
		astSchema,
		make(map[string]*graphql.Object),
		make(map[string]*graphql.Interface),
		make(map[string]*graphql.Union),
		make(map[string]*graphql.Enum),
		make(map[string]*graphql.InputObject),
	}

	schema, err := graphql.NewSchema(graphql.SchemaConfig{
		Query:    convertSchemaObject(astSchema.Query, schemaCtx),
		Mutation: convertSchemaObject(astSchema.Mutation, schemaCtx),
	})

	schemaPtr = new(graphql.Schema)
	*schemaPtr = schema

	return
}
