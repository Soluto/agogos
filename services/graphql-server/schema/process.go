package schema

import (
	"agogos/extensions/upstreams"
	"agogos/extensions/upstreams/authentication"
	agogos "agogos/generated"
	"agogos/server"

	"github.com/graphql-go/graphql"
	log "github.com/sirupsen/logrus"
	"github.com/vektah/gqlparser"
	"github.com/vektah/gqlparser/ast"
)

func parseSdl(name, sdl string) (*ast.Schema, error) {
	schema, err := gqlparser.LoadSchema(&ast.Source{
		Name:  name,
		Input: sdl,
	})

	if err != nil {
		return nil, err
	}

	return schema, nil
}

func transformToSchema(config *agogos.ConfigurationMessage) (*graphql.Schema, error) {
	astSchema, err := parseSdl(
		"schema registry sdl",
		config.Schema.Definition,
	)
	if err != nil {
		log.WithError(err).WithField("schema", config.Schema.Definition).Warn("Error parsing SDL")
		return nil, err
	}

	upstreamsMap := createUpstreams(config.Upstreams, config.UpstreamAuthCredentials)
	serverContext := server.CreateServerContext(upstreamsMap)

	convertedSchema, err := convertSchema(serverContext, astSchema)
	if err != nil {
		log.WithError(err).Warn("Error converting schema")
		return nil, err
	}

	return convertedSchema, nil
}

func createUpstreams(upsConfigs []*agogos.Upstream, upsAuthConfigs []*agogos.UpstreamAuthCredentials) map[string]upstreams.Upstream {
	upsAuths := make(map[string]map[string]authentication.UpstreamAuthentication)
	for _, upsAuthConfig := range upsAuthConfigs {
		if _, ok := upsAuths[upsAuthConfig.AuthType]; !ok {
			upsAuths[upsAuthConfig.AuthType] = make(map[string]authentication.UpstreamAuthentication)
		}

		upsAuths[upsAuthConfig.AuthType][upsAuthConfig.Authority] = authentication.CreateFromConfig(upsAuthConfig)
	}

	ups := make(map[string]upstreams.Upstream, len(upsConfigs))

	for _, upConfig := range upsConfigs {
		var matchedUpsAuth authentication.UpstreamAuthentication
		if m, ok := upsAuths[upConfig.Auth.AuthType]; ok {
			if upsAuth, ok := m[upConfig.Auth.Authority]; ok {
				matchedUpsAuth = upsAuth
			}
		}

		ups[upConfig.Host] = upstreams.CreateFromConfig(upConfig, matchedUpsAuth)
	}

	return ups
}

type SchemaResult struct {
	Schema *graphql.Schema
	Error  error
}

func Process(configCh chan *agogos.ConfigurationMessage) chan SchemaResult {
	schemaCh := make(chan SchemaResult)

	go func() {
		defer close(schemaCh)

		for {
			config := <-configCh
			log.Infoln("Got new configuration from registry")
			schema, err := transformToSchema(config)
			schemaCh <- SchemaResult{
				Schema: schema,
				Error:  err,
			}
		}
	}()

	return schemaCh
}
