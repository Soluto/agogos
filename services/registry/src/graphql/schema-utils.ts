import {
    DefinitionNode,
    DocumentNode,
    FieldDefinitionNode,
    InputValueDefinitionNode
} from "graphql";
import {
    clone,
    equals,
    flatten,
    groupBy,
    indexBy,
    map,
    pipe,
    unionWith,
    uniq,
    values
} from "ramda";

function validateSameFieldDefs(
    fieldDefA: FieldDefinitionNode | InputValueDefinitionNode,
    fieldDefB: FieldDefinitionNode | InputValueDefinitionNode
) {
    const { directives: _, ...fieldDefWithNoDirectivesA } = fieldDefA;
    const { directives: __, ...fieldDefWithNoDirectivesB } = fieldDefB;
    if (!equals(fieldDefWithNoDirectivesA, fieldDefWithNoDirectivesB)) {
        throw new Error(`Same field definition should be identical except for directives - field ${
            fieldDefA.name.value
            }`);
    }
}

function mergeSameFieldDefinitions(
    fieldDefinitions: ReadonlyArray<
        FieldDefinitionNode | InputValueDefinitionNode
    >
): ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode> {
    if (fieldDefinitions.length === 0) { return []; }
    if (fieldDefinitions.length === 1) { return fieldDefinitions; }

    const [fieldDefA, fieldDefB, ...rest] = fieldDefinitions;
    validateSameFieldDefs(fieldDefA, fieldDefB);

    const { directives: _, ...otherFields } = fieldDefA;
    const fieldWithMergedDirectives = {
        ...otherFields,
        directives: uniq([
            ...(fieldDefA.directives || []),
            ...(fieldDefB.directives || [])
        ])
    };

    return mergeSameFieldDefinitions([fieldWithMergedDirectives, ...rest]);
}

const mergeFields = (
    l: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>,
    r: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>
) =>
    pipe(
        groupBy(
            (fieldDef: FieldDefinitionNode | InputValueDefinitionNode) =>
                fieldDef.name.value
        ),
        values,
        map(mergeSameFieldDefinitions),
        flatten
    )([...l, ...r]);

export function mergeDocuments(schemas: DocumentNode[]) {
    return clone(schemas).reduce((acc: DocumentNode, next: DocumentNode) => {
        const index = indexBy((x: DefinitionNode) => {
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "UnionTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                return `${x.kind}_${x.name.value}`;
            }
            return "";
        })(acc.definitions);

        const get = (kind: string, value: string) => index[`${kind}_${value}`];

        next.definitions.forEach(x => {
            if (
                (x.kind !== "ObjectTypeDefinition" &&
                    x.kind !== "UnionTypeDefinition" &&
                    x.kind !== "InputObjectTypeDefinition") ||
                !get(x.kind, x.name.value)
            ) {
                (acc.definitions as any).push(x);
                return;
            }

            const ref = get(x.kind, x.name.value);
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                (ref as any).directives = uniq([
                    ...(ref as any).directives,
                    ...(x.directives || [])
                ]);
                (ref as any).fields = mergeFields((ref as any).fields, x.fields || []);

                if (x.kind === "ObjectTypeDefinition") {
                    (ref as any).interfaces = uniq([
                        ...(ref as any).interfaces,
                        ...(x.interfaces || [])
                    ]);
                }
                return;
            }

            if (x.kind === "UnionTypeDefinition") {
                (ref as any).types = unionWith(equals, (ref as any).types, x.types || []);
                return;
            }
        });

        return acc;
    });
}
