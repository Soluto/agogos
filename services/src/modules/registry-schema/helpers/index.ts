import proceedPolicy from './policy-helper';

export { proceedPolicy };
