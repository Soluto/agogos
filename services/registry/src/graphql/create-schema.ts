import gql from "graphql-tag";
import { mergeDocuments } from "./schema-utils";
import { DocumentNode } from "graphql";

type GqlSource = {
  source: string;
  gql: string;
};

export type GqlSources = { [source: string]: string };

function parseGqlSource(gqlSource: GqlSource) {
  try {
    return gql(gqlSource.gql);
  } catch (e) {
    throw {
      message: `Failed to load schema for source -  ${gqlSource.source} - ${
        e.message
      }`,
      schema: gqlSource,
      error: e
    };
  }
}

function mergeAllDocuments(docs: DocumentNode[]) {
  try {
    return mergeDocuments(docs);
  } catch (e) {
    throw {
      message: `Failed to merge schemas - ${e.message}`,
      schemas: docs,
      error: e
    };
  }
}

export const makeGqlDocumentFromGqlSources = (gqlSources: GqlSources) => {
  const documentNodes = Object.entries(gqlSources).map(([source, gql]) =>
    parseGqlSource({ source, gql })
  );
  return mergeAllDocuments(documentNodes);
};
