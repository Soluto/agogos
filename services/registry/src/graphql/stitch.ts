import {DocumentNode, parse} from 'graphql';
import {mergeDocuments} from './schema-utils';
import {AggObjsByName} from '../sync/sync-service';
import {SchemaConfig} from '../sync/object-types';

function parseGqlSource(schemaName: string, schema: string) {
    try {
        return parse(schema);
    } catch (e) {
        throw {
            error: e,
            message: `Failed to load schema for source -  ${schemaName} - ${e.message}`,
            schema: {schemaName, schema},
        };
    }
}

function mergeAllDocuments(docs: DocumentNode[]) {
    try {
        return mergeDocuments(docs);
    } catch (e) {
        throw {
            error: e,
            message: `Failed to merge schemas - ${e.message}`,
            schemas: docs,
        };
    }
}

export const stitchSchemas = (gqlSchemas: AggObjsByName) => {
    const documentNodes = Object.entries(gqlSchemas).map(([schemaName, schema]: [string, SchemaConfig]) =>
        parseGqlSource(schemaName, schema.definition)
    );
    return mergeAllDocuments(documentNodes);
};
