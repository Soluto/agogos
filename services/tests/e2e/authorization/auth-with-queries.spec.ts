import { print } from 'graphql';
import { GraphQLClient } from 'graphql-request';
import { gql } from 'apollo-server-core';
import {
  createSchemaMutation,
  CreatePolicyMutationResponse,
  createPolicyMutation,
  UpdateSchemasMutationResponse,
  emptySchema,
} from '../../helpers/registry-request-builder';
import { sleep } from '../../helpers/utility';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { getToken } from '../../helpers/get-token';
import { schema, policies, AllowedEmployeeQueryResponse, employeeQuery } from './auth-with-queries.schema';

describe('Authorization with queries', () => {
  let gatewayClient: GraphQLClient;
  let registryClient: GraphQLClient;

  beforeAll(async () => {
    gatewayClient = new GraphQLClient('http://localhost:8080/graphql');
    registryClient = new GraphQLClient('http://localhost:8090/graphql');

    const accessToken = await getToken();
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);
    expect.addSnapshotSerializer(GraphQLErrorSerializer);
  });

  afterAll(async () => {
    const schemaResponse: UpdateSchemasMutationResponse = await registryClient.request(createSchemaMutation, {
      schema: emptySchema(schema),
    });
    expect(schemaResponse.updateSchemas.success).toBeTruthy();

    // Wait for gateway to update before next tests
    await sleep(Number(process.env.WAIT_FOR_REFRESH_ON_GATEWAY) | 1500);
  });

  test('Setup policies', async () => {
    const policyResponse: CreatePolicyMutationResponse = await registryClient.request(createPolicyMutation, {
      policies,
    });
    expect(policyResponse.updatePolicies.success).toBeTruthy();

    const schemaResponse: UpdateSchemasMutationResponse = await registryClient.request(createSchemaMutation, {
      schema,
    });
    expect(schemaResponse.updateSchemas.success).toBeTruthy();

    // Wait for gateway to update before next tests
    await sleep(Number(process.env.WAIT_FOR_REFRESH_ON_GATEWAY) | 1500);
  });

  test('Query always-allow policy', async () => {
    const response = await gatewayClient.request(
      print(gql`
        query {
          policy {
            auth_with_query___always_allow {
              allow
            }
          }
        }
      `)
    );
    expect(response).toMatchSnapshot();
  });

  test('Query alwaysDeny policy', async () => {
    let response;
    try {
      response = await gatewayClient.request(
        print(gql`
          query {
            policy {
              auth_with_query___always_allow {
                allow
              }
            }
          }
        `)
      );
    } catch (e) {
      response = e.response;
    }
    expect(response).toMatchSnapshot();
  });

  test('Query protected field', async () => {
    let response;
    try {
      response = await gatewayClient.request(
        print(gql`
          query {
            classifiedDepartments {
              id
              name
            }
          }
        `)
      );
    } catch (e) {
      response = e.response;
    }
    expect(response).toMatchSnapshot();
  });

  test('Query allowed employee', async () => {
    const response: AllowedEmployeeQueryResponse = await gatewayClient.request(employeeQuery('allowedEmployee'));
    expect(response.allowedEmployee).toMatchSnapshot();
  });

  test('Query denied employee 1', async () => {
    let response;
    try {
      response = await gatewayClient.request(employeeQuery('deniedEmployee1'));
    } catch (e) {
      response = e.response;
    }
    expect(response).toMatchSnapshot();
  });

  test('Query denied employee 2', async () => {
    let response;
    try {
      response = await gatewayClient.request(employeeQuery('deniedEmployee2'));
    } catch (e) {
      response = e.response;
    }
    expect(response).toMatchSnapshot();
  });
});
