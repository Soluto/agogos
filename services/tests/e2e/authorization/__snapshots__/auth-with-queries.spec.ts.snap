// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization with queries Query allowed employee 1`] = `
Object {
  "address": "Tel Aviv",
  "id": "1",
  "name": "John Smith",
}
`;

exports[`Authorization with queries Query alwaysAllow policy 1`] = `
Object {
  "policy": Object {
    "auth_with_query___alwaysAllow": Object {
      "allow": true,
    },
  },
}
`;

exports[`Authorization with queries Query alwaysDeny policy 1`] = `
Object {
  "policy": Object {
    "auth_with_query___alwaysAllow": Object {
      "allow": true,
    },
  },
}
`;

exports[`Authorization with queries Query denied employee 1 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy notClassified in namespace auth_with_query",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "deniedEmployee1": {
      "id": "2",
      "name": "Mark Zuckerberg",
      "address": null
    }
  }
}
`;

exports[`Authorization with queries Query denied employee 2 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy notClassified in namespace auth_with_query",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "deniedEmployee2": {
      "id": "2",
      "name": "Tom Baker",
      "address": null
    }
  }
}
`;

exports[`Authorization with queries Query protected field 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy alwaysDenied in namespace auth_with_query",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;
